src/api/axios/ApiErrorStrategy.ts

```diff
@@ -0,0 +1,44 @@
+import { AxiosError } from 'axios';
+
+import { APIErrorOutput, ApiErrorHandlerOptions, ErrorHandlingStrategy } from 'api/types/types';
+
+export class ErrorResponseStrategy implements ErrorHandlingStrategy {
+  getErrorObject(error: AxiosError) {
+    return error.response;
+  }
+}
+
+export class ErrorDataStrategy implements ErrorHandlingStrategy {
+  getErrorObject(error: AxiosError) {
+    return error.response?.data;
+  }
+}
+
+export class ErrorDetailsStrategy implements ErrorHandlingStrategy {
+  getErrorObject(error: AxiosError<{ error: unknown }>) {
+    return error.response?.data?.error;
+  }
+}
+
+export class ApiErrorHandler {
+  private strategy: ErrorHandlingStrategy;
+  private defaultErrorStatus: ApiErrorHandlerOptions['defaultErrorStatus'];
+
+  constructor(strategy: ErrorHandlingStrategy) {
+    this.strategy = strategy;
+    this.defaultErrorStatus = 500;
+  }
+
+  narrowErrorData(error: AxiosError): APIErrorOutput {
+    return {
+      status: error.status || this.defaultErrorStatus,
+      code: error.code,
+      message: error.message,
+    };
+  }
+
+  handleError(error: AxiosError): APIErrorOutput {
+    const errorData = this.strategy.getErrorObject(error);
+    return this.narrowErrorData(errorData || error);
```

Comment:
As the `narrowErrorData` function expects `AxiosError` maybe it will be better if the strategy also returns the same type? It may be misleading that it may return `any` but then some data can be lost during narrowing.

Reply:
yup, the strategy should return a known type, it can accept axios error but the real purpose should be to map the error body to a known output type. Default strategy should expect error object as we discussed

Reply:
@pzmuda1 @qh4r I've added few improvements. Please let me know if i misunderstood something.